src/matrix.h

```diff
@@ -20,96 +21,465 @@
+	const float radians = -Radians(degrees);
+	const float c = cosf(radians);
+	const float s = sinf(radians);
+	
+	return Mat4_FromColumns((const float[]) {
+		axis.x * axis.x + c * (1.f - axis.x * axis.x),
+		axis.x * axis.y * (1.f - c) + axis.z * s,
+		axis.z * axis.x * (1.f - c) - axis.y * s,
+		0.f
+	}, (const float[]) {
+		axis.x * axis.y * (1.f - c) - axis.z * s,
+		axis.y * axis.y + c * (1.f - axis.y * axis.y),
+		axis.y * axis.z * (1.f - c) + axis.x * s,
+		0.f
+	}, (const float[]) {
+		axis.z * axis.x * (1.f - c) + axis.y * s,
+		axis.y * axis.z * (1.f - c) - axis.x * s,
+		axis.z * axis.z + c * (1.f - axis.z * axis.z),
+		0.f
+	}, (const float[]) {
+		0.f,
+		0.f,
+		0.f,
+		1.f
+	});
+}
+
+/**
+ * @return A scale matrix.
+ */
+static inline mat4_t __attribute__ ((warn_unused_result)) Mat4_FromScale3(const vec3_t scale) {
+	return Mat4((const float []) {
+		scale.x, 0.f,     0.f,     0.f,
+		0.f,     scale.y, 0.f,     0.f,
+		0.f,     0.f,     scale.z, 0.f,
+		0.f,     0.f,     0.f,     1.f
+	});
+}
+
+/**
+ * @return A scale matrix.
+ */
+static inline mat4_t __attribute__ ((warn_unused_result)) Mat4_FromScale(const float scale) {
+	return Mat4_FromScale3(Vec3(scale, scale, scale));
+}
+
+/**
+ * @brief Fetch the three directional vectors and/or translation from this matrix.
+ */
+static inline void Mat4_ToVectors(const mat4_t in, vec3_t *forward, vec3_t *right, vec3_t *up, vec3_t *translation) {
+
+	if (forward) {
+		*forward = Vec3(in.m[0][0], in.m[0][1], in.m[0][2]);
+	}
+
+	if (right) {
+		*right = Vec3(in.m[1][0], in.m[1][1], in.m[1][2]);
+	}
+
+	if (up) {
+		*up = Vec3(in.m[2][0], in.m[2][1], in.m[2][2]);
+	}
+
+	if (translation) {
+		*translation = Vec3(in.m[3][0], in.m[3][1], in.m[3][2]);
+	}
+}
+
+/**
+ * @return A matrix defined by the three directional vectors and translation vectors.
+ */
+static inline mat4_t __attribute__ ((warn_unused_result)) Mat4_FromVectors(const vec3_t forward, const vec3_t right, const vec3_t up, const vec3_t translation) {
+	return Mat4_FromColumns(
+		(const float[]) { forward.x, right.x, up.x, translation.x },
+		(const float[]) { forward.y, right.y, up.y, translation.y },
+		(const float[]) { forward.z, right.z, up.z, translation.z },
+		(const float[]) { 0.f, 0.f, 0.f, 1.f }
+	);
+}
+
+/**
+ * @return The input vector transformed by the specified matrix.
+ */
+static inline vec3_t __attribute__ ((warn_unused_result)) Mat4_Transform(const mat4_t m, const vec3_t v) {
+	return Vec3(
+		v.x * m.m[0][0] + v.y * m.m[1][0] + v.z * m.m[2][0] + m.m[3][0],
+		v.x * m.m[0][1] + v.y * m.m[1][1] + v.z * m.m[2][1] + m.m[3][1],
+		v.x * m.m[0][2] + v.y * m.m[1][2] + v.z * m.m[2][2] + m.m[3][2]
+	);
+}
+
+/**
+ * @return The transformed positive distance plane (A*x+B*y+C*z-D=0).
+ */
+static inline vec4_t __attribute__ ((warn_unused_result)) Mat4_TransformQuakePlane(const mat4_t in, const vec3_t n, float d) {
```

Comment:
We can probably just call this Mat4_TransformPlane.

Reply:
And make it take in a double for the distance, so that Quemap can use it.

Reply:
Quemap doesn't use this function at all, so it seems to be okay to keep to float. Does it have its own version maybe?

Reply:
Ah, no, I guess it doesn't need this. Makes sense... Quemap lights inline models in their natural (0, 0, 0) positions, so no transforms are required to accurately do lighting traces.